open_lm/datapreprocess/ray/tokenize_shuffle.py

```diff
@@ -571,23 +647,27 @@ def main(args):
     else:
         ds = ds.repartition(num_cores * num_nodes, shuffle=False)
     counter = GlobalCounter.remote()
-    ds = ds.map_batches(
-        map_write_wds,
-        batch_size=wds_chunk_size,
+    out_folder = args.output.rstrip("/")
+    # first map buffer_write over rows, it will create an actor (which hopefully will be scheduled locally)
+    write_status = ds.map(
+        buffer_write,
         fn_kwargs={
-            "batch_size": wds_chunk_size,
-            "folder": args.output.rstrip("/"),
+            "folder": out_folder,
             "counter": counter,
+            "buffer_size": args.wds_chunk_size,
+            "num_writers_per_node": num_writers_per_node,
         },
-        batch_format="pandas",
-    )
-
-    # Sort by shard name
-    ds = ds.repartition(1)
-    ds = ds.sort(key="shard")
-    jsonl_lines = ds.take_all()
-    write_manifest(jsonl_lines, args)
-
+    ).take_all()
+    # after the write is done, grab all actors of class BufferedShardWriter
+    buffer_writers = [ray.get_actor(x.name) for x in list_actors(filters=[("class_name", "=", "BufferedShardWriter")])]
+    # flush the remaining buffers, this should be the *only* shards that are less than wds_chunk_size
+    flushed_buffers = [bw._flush_buffer.remote(out_folder, counter) for bw in buffer_writers]
```

Comment:
This will also assure that the small shards are only the last ones, right?
Would it make sense to just drop them too?

Reply:
no lets just keep them its fine, by construction the last shards on each writer actor will be the small one